notarization_poller/src/notarization_poller/task.py

```diff
@@ -0,0 +1,321 @@
+from scriptworker_client.exceptions import Download404, DownloadError, TaskError
+from scriptworker_client.utils import load_json_or_yaml, makedirs, rm, run_command
+
+log = logging.getLogger(__name__)
+NOTARIZATION_POLL_REGEX = re.compile(r"Status: (?P<status>success|invalid)")
+
+
+class Task:
+    """Manages all information related to a single running task."""
+
+    reclaim_fut = None
+    task_fut = None
+    complete = False
+    process = None
+    uuids = None
+
+    def __init__(self, config, claim_task, event_loop=None):
+        """Initialize Task."""
+        self.config = config
+        self.task_id = claim_task["status"]["taskId"]
+        self.run_id = claim_task["runId"]
+        self.claim_task = claim_task
+        self.event_loop = event_loop or asyncio.get_event_loop()
+        self.task_dir = os.path.join(config["work_dir"], self.task_id)
+        self.log_path = os.path.join(self.task_dir, "live_backing.log")
+        self.poll_log_path = os.path.join(self.task_dir, "polling.log")
+        rm(self.task_dir)
+        makedirs(self.task_dir)
+        self.reclaim_fut = event_loop.create_task(self.reclaim_task())
+        self._reclaim_task = {}
+        self.task_fut = event_loop.create_task(self.run_task())
+
+    @property
+    def task_credentials(self):
+        """Return the temporary credentials returned from [re]claimWork."""
+        return self._reclaim_task.get("credentials", self.claim_task["credentials"])
+
+    async def reclaim_task(self):
+        """Try to reclaim a task from the queue.
+
+        This is a keepalive.  Without it the task will expire and be re-queued.
+
+        A 409 status means the task has been resolved.
+
+        Raises:
+            TaskclusterRestFailure: on non-409 status_code from
+                `taskcluster.aio.Queue.reclaimTask`
+
+        """
+        while True:
+            log.debug("waiting %s seconds before reclaiming..." % self.config["reclaim_interval"])
+            await asyncio.sleep(self.config["reclaim_interval"])
+            log.debug("Reclaiming task %s %s", self.task_id, self.run_id)
+            try:
+                async with aiohttp.ClientSession() as session:
+                    temp_queue = Queue(options={"credentials": self.task_credentials, "rootUrl": self.config["taskcluster_root_url"]}, session=session)
+                    self.reclaim_task = await temp_queue.reclaimTask(self.task_id, self.run_id)
+            except taskcluster.exceptions.TaskclusterRestFailure as exc:
+                if exc.status_code == 409:
+                    log.warning("Stopping task %s %s after receiving 409 response from reclaim_task: %s %s", self.task_id, self.run_id)
+                    exit_status = STATUSES["superseded"]
+                else:
+                    log.exception("reclaim_task unexpected exception: %s %s", self.task_id, self.run_id)
+                    exit_status = STATUSES["internal-error"]
+                await self.stop(status=exit_status)
```

Comment:
i *think* it makes sense to instead cancel the `run_task` future here, and handle `CancelledError` there to handle uploading logs, and resolving the task.

Reply:
+1.